js/webtorrent/entry.js

```diff
@@ -1,148 +1,186 @@
 const parseTorrent = require('parse-torrent')
+const path = require('path')
+const querystring = require('querystring')
 const React = require('react')
 const ReactDOM = require('react-dom')
+const url = require('url')
 const WebTorrentRemoteClient = require('webtorrent-remote/client')
 
-// React Components
-const MediaViewer = require('./components/mediaViewer')
-const TorrentViewer = require('./components/torrentViewer')
+const App = require('./components/app')
+const messages = require('../constants/messages')
 
 // Stylesheets
 require('../../less/webtorrent.less')
 require('../../node_modules/font-awesome/css/font-awesome.css')
 
-// UI state object. Pure function from `state` -> React element.
-const state = {
-  torrentID: window.decodeURIComponent(window.location.hash.substring(1)),
-  parsedTorrent: null,
-  client: null,
+// Pure function from state -> React elements.
+const store = {
+  name: null, // Torrent name
+  ix: null, // Selected file index
+  torrentId: window.decodeURIComponent(window.location.hash.substring(1)),
+  torrentIdProtocol: null,
   torrent: null,
+  serverUrl: null,
   errorMessage: null
 }
-window.state = state /* for easier debugging */
 
-state.parsedTorrent = parseTorrent(state.torrentID)
+let client, server
+
+init()
+
+function init () {
+  const parsedUrl = url.parse(store.torrentId)
+  store.torrentIdProtocol = parsedUrl.protocol
+
+  // `ix` param can be set by query param or hash, e.g. ?ix=1 or #ix=1
+  if (parsedUrl.protocol === 'http:' || parsedUrl.protocol === 'https:') {
+    store.name = path.basename(parsedUrl.pathname)
+    store.ix = getIxQuery(parsedUrl)
+    if (store.ix == null) store.ix = getIxHash(parsedUrl)
+  } else if (parsedUrl.protocol === 'magnet:') {
+    let parsedTorrent
+    try {
+      parsedTorrent = parseTorrent(store.torrentId)
+    } catch (err) {
+      return onError(new Error('Invalid torrent identifier'))
+    }
+    store.name = parsedTorrent.name
+    store.ix = parsedTorrent.ix
+    if (store.ix == null) store.ix = getIxHash(parsedUrl)
+  } else {
+    return onError(new Error('Invalid torrent identifier'))
+  }
+
+  // Create the client, set up IPC to the WebTorrentRemoteServer
+  client = new WebTorrentRemoteClient(send)
+  client.on('warning', onWarning)
+  client.on('error', onError)
+
+  ipc.on(messages.TORRENT_MESSAGE, function (e, msg) {
+    client.receive(msg)
+  })
+
+  // Check whether we're already part of this swarm. If not, show a Start button.
+  client.get(store.torrentId, function (err, torrent) {
+    if (!err) {
+      store.torrent = torrent
+      initTorrent(torrent)
+    }
+    update()
+  })
+
+  // Clean up the client. Note: Since this does IPC, it's not guaranteed to send
+  // before the page is closed. But that's okay; webtorrent-remote expects regular
+  // heartbeats and assumes clients are dead after 30s without one. So basically,
+  // this is an optimization to destroy the client sooner.
```

Comment:
Added a much more descriptive comment here. We're not relying on this IPC succeeding before the page is unloaded. (It only succeeds like 50% of the time.)